src/Microsoft.AspNetCore.Hosting/Internal/WebHostOptions.cs

```diff
@@ -22,6 +22,7 @@ public WebHostOptions(IConfiguration configuration)
             }
 
             Application = configuration[WebHostDefaults.ApplicationKey];
+            EnvironmentApplication = configuration[WebHostDefaults.EnvironmentApplicationKey];
```

Comment:
I can also just default the `ApplicationName` of the `IApplicationEnvironment` service to what is set by `Application`. Note that only `UseStartup(string)` or configuration will set this variable not `UseStartup(Type)` or `UseStartup<Type>()`.